dag.go

```diff
@@ -102,6 +109,152 @@ func (m *Manifest) RootCID() cid.Cid {
 // 	return nodes
 // }
 
+// SubDAG returns a Manifest of a DAG at a specific hash in an already existing Manifest
+func SubDAG(m *Manifest, id string) (*Manifest, error) {
+	info := &Info{
+		Manifest: m,
+	}
+	root, err := m.IDIndex(id)
+	if err != nil {
+		return nil, err
+	}
+	converter, err := newSubDAGConverter(info, root)
+	if err != nil {
+		return nil, err
+	}
+	converter.Convert()
+	return converter.Manifest, nil
+}
+
+func newSubDAGConverter(prevInfo *Info, root int) (*subDAGConverter, error) {
+	if prevInfo.Manifest == nil {
+		return nil, fmt.Errorf("no manifest provided")
+	}
+	if root < 0 || root > len(prevInfo.Manifest.Nodes)-1 {
+		return nil, ErrIndexOutOfRange
+	}
+	m := prevInfo.Manifest
+	PrevLabels := map[int]string{}
+	for path, index := range prevInfo.Labels {
+		PrevLabels[index] = path
+	}
+
+	lastLink := m.Links[len(m.Links)-1]
+	return &subDAGConverter{
+		Conversion:         map[int]int{},
+		PrevManifest:       m,
+		Manifest:           &Manifest{Nodes: []string{}, Links: [][2]int{}},
+		currentParentIndex: 0,
+		Parents:            []int{root},
+		LastBranchIndex:    lastLink[0],
+		PrevSizes:          prevInfo.Sizes,
+		Sizes:              []uint64{},
+		PrevLabels:         PrevLabels,
+		Labels:             map[string]int{},
+	}, nil
+}
+
+type subDAGConverter struct {
+	Conversion         map[int]int
+	PrevManifest       *Manifest
+	currentParentIndex int
+	Parents            []int
+	LastBranchIndex    int
+	Manifest           *Manifest
+	PrevSizes          []uint64
+	Sizes              []uint64
+	PrevLabels         map[int]string
+	Labels             map[string]int
```

Comment:
I find these two field names confusing. Especially since they are similar sounding, but have opposite types. Maybe I'm wrong, but it seems like `PrevLabels` is not really a "previous" something, but more like an inverter of the labels. If so, how about naming them `LabelsMap` and `InverseLabelsMap`?

Reply:
Yes you are right this makes no sense. `Labels` is exactly what ends up in the new sub-dag.Info's `Labels` field, tho. So I think maybe `InverseLabels` and `Labels` might be the most accurate.